### PR TITLE
Partitions

### LINKED ISSUE #788: FEAT: move table partitions to the "structure" view, under a sub-tab "partitions"
I got a lot of partition tables, I really do not want to see them in the tree view.

Can you guys hide them under the origin table and would be better to create a folder "partitions"

Screen Shot 2021-09-03 at 2 59 45 pm

### PR DESCRIPTION
**Pull request by not-night-but**
Close #788 

Allow Users to manage partitions in table structure view.

Adds a new tab in the structure view that only pops up in postgres if the table is a partitioned table:
image

This shows the name of the partition and the partitioning expression.

This view also allows users to add a partition, detach existing partitions, and edit the partitioning expression. These changes can either be directly applied, or copied to SQL in a new tab.
image

Setup SQL for testing:
```sql
CREATE TABLE
  employee (
    recordid serial NOT NULL,
    stuff character varying(255) NULL,
    age integer NULL
  ) PARTITION BY RANGE (age);
  
CREATE TABLE employee_young PARTITION OF employee
FOR VALUES FROM (18) to (25)

CREATE TABLE employee_mid PARTITION OF employee
FOR VALUES FROM (26) to (40)

CREATE TABLE employee_old PARTITION OF employee
FOR VALUES FROM (41) to (55)

insert into "employee" ("age", "recordid", "stuff") values (23, 4, 'Day');
insert into "employee" ("age", "recordid", "stuff") values (20, 5, 'Mars');
insert into "employee" ("age", "recordid", "stuff") values (23, 11, 'Kolsen');
insert into "employee" ("age", "recordid", "stuff") values (20, 12, 'Ellie');
insert into "employee" ("age", "recordid", "stuff") values (28, 1, 'Bob');
insert into "employee" ("age", "recordid", "stuff") values (34, 3, 'Teresa');
insert into "employee" ("age", "recordid", "stuff") values (28, 8, 'John');
insert into "employee" ("age", "recordid", "stuff") values (34, 10, 'Luna');
insert into "employee" ("age", "recordid", "stuff") values (53, 2, 'Jane');
insert into "employee" ("age", "recordid", "stuff") values (48, 6, 'Shane');
insert into "employee" ("age", "recordid", "stuff") values (49, 7, 'Patrick');
insert into "employee" ("age", "recordid", "stuff") values (53, 9, 'Wayne');
insert into "employee" ("age", "recordid", "stuff") values (48, 13, 'Chelsea');
insert into "employee" ("age", "recordid", "stuff") values (49, 14, 'ELla');
```

### CODE REVIEW

File: apps/studio/src/components/TabTableProperties.vue [hunk 1]
```diff
@@ -238,9 +255,11 @@ export default {
       // this.properties = null
       try {
         await this.$store.dispatch('updateTableColumns', this.table)
+        await this.$store.dispatch('updateTablePartitions', this.table)
         this.primaryKeys = await this.connection.getPrimaryKeys(this.table.name, this.table.schema)
         if (this.table.entityType === 'table') {
           this.properties = await this.connection.getTableProperties(this.table.name, this.table.schema)
+          console.log('Table Properties: ', this.properties)
```
Comment (not-night-but) [inline apps/studio/src/components/TabTableProperties.vue:262]: Remove log

File: apps/studio/src/lib/db/client.ts [hunk 2]
```diff
@@ -104,8 +108,7 @@ export interface IDbConnectionServerConfig {
   port: Nullable<number>,
   domain: Nullable<string>,
   socketPath: Nullable<string>,
-  socketPathEnabled: boolean,
-  user: Nullable<string>,
+  socketPathEnabled: boolean, user: Nullable<string>,
```
Comment (not-night-but) [inline apps/studio/src/lib/db/client.ts:111]: ????

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 3]
```diff
@@ -156,7 +156,7 @@ export default async function (server: any, database: any): Promise<DatabaseClie
 
   const version = await getVersion(conn)
 
-  const features = version.isRedshift ? { customRoutines: true, comments: false, properties: false } : { customRoutines: true, comments: true, properties: true}
+  const features = version.isRedshift ? { customRoutines: true, comments: false, properties: false, partitions: false } : { customRoutines: true, comments: true, properties: true, partitions: true}
```
Comment (not-night-but) [inline apps/studio/src/lib/db/clients/postgresql.ts:159]: Should probably check to see if partitions are supported on for the version of Postgres running, as it was only introduced in version 10 I believe.

**Comment by not-night-but** [comment]
@rathboma I think I'm pretty much done with this. Let me know if there's anything you want me to change!

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 4]
```diff
@@ -229,21 +233,47 @@ export function disconnect(conn: HasPool) {
 
 export async function listTables(conn: HasPool, filter: FilterOptions = { schema: 'public' }) {
   const schemaFilter = buildSchemaFilter(filter, 'table_schema');
+  // @Day: selecting all tables that are not partitions
```
Comment (not-night-but) [inline apps/studio/src/lib/db/clients/postgresql.ts:236]: We should only do this if the db supports partitions

**Comment by rathboma** [comment]
Just keep this pr scoped to Postgres for now. Other dabs can be in a follow up pr.

let me know when ready for rereview

File: apps/studio/src/components/TabTableProperties.vue [hunk 5]
```diff
@@ -190,6 +203,10 @@ export default {
         if (p.needsProperties && !this.connection.supportedFeatures().properties) {
           return false
         }
+
+        if (p.needsPartitions && (!this.connection.supportedFeatures().partitions || this.table.tabletype !== 'p')) {
```
Comment (not-night-but) [inline apps/studio/src/components/TabTableProperties.vue]: magic value :facepalm:

**Comment by not-night-but** [comment]
@rathboma alrighty i believe this is ready for review.

File: apps/studio/src/lib/db/client.ts [hunk 6]
```diff
@@ -316,6 +324,11 @@ function selectTopStream(
   return database.connection?.selectTopStream(database.database, table, orderBy, filters, chunkSize, schema)
 }
 
+function listTablePartitions(server: IDbConnectionServer, database: IDbConnectionDatabase, table?: string) {
+  checkIsConnected(server, database);
+  return database.connection?.listTablePartitions(table);
```
Comment (rathboma) [inline apps/studio/src/lib/db/client.ts:329]: Why not use bindAsync here?

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 7]
```diff
@@ -228,21 +234,66 @@ export function disconnect(conn: HasPool) {
...
+  const sql = `
+    SELECT 
+      ps.schemaname AS schema,
+      ps.relname AS name,
+      pg_get_expr(pt.relpartbound, pt.oid, true) AS expression
+    FROM pg_class base_tb
+      JOIN pg_inherits i ON i.inhparent = base_tb.oid
+      JOIN pg_class pt ON pt.oid = i.inhrelid
+      JOIN pg_stat_all_tables ps ON ps.relid = i.inhrelid
+    WHERE base_tb.oid = '${tableName}'::regclass
```
Comment (rathboma) [inline apps/studio/src/lib/db/clients/postgresql.ts:288]: Need to escape tableName. Use the string wrap

File: apps/studio/src/store/index.ts [hunk 8]
```diff
@@ -422,6 +422,19 @@ const store = new Vuex.Store<State>({
       }
     },
 
+    async updateTablePartitions(context, table: TableOrView) {
```
Comment (rathboma) [inline apps/studio/src/store/index.ts:425]: So we don’t do this with indexes or relations or triggers. You should emulate how those properties tabs work so we’re consistent

File: apps/studio/src/store/index.ts [hunk 9]
```diff
@@ -435,6 +448,9 @@ const store = new Vuex.Store<State>({
           onlyTables.forEach((t) => {
             t.entityType = 'table'
             t.columns = []
+            t.partitions = []
+            if (context.state.connection.connectionType != 'postgresql')
```
Comment (rathboma) [inline apps/studio/src/store/index.ts:452]: I don’t like this connection type awareness living here. Can we check supportedFeatures or have listTables set this to mull?

File: shared/src/lib/sql/change_builder/PostgresqlChangeBuilder.ts [hunk 10]
```diff
@@ -11,4 +11,50 @@ export class PostgresqlChangeBuilder extends ChangeBuilderBase {
   wrapIdentifier = wI
   wrapLiteral = wL
   escapeString = wrapString
+
+  createPartition(spec: PartitionItem) {
+    const result = `
```
Comment (rathboma) [inline shared/src/lib/sql/change_builder/PostgresqlChangeBuilder.ts:16]: Escape escape escape!

File: shared/src/lib/sql/change_builder/PostgresqlChangeBuilder.ts [hunk 11]
```diff
@@ -11,4 +11,50 @@ export class PostgresqlChangeBuilder extends ChangeBuilderBase {
...
+    if (!partitions?.length) return null;
+    return partitions.map((part) => this.detachPartition(part)).join(';');
+  }
+
+  alterPartition(alter: PartitionExpressionChange) {
+    const detachPartition = this.detachPartition(alter.partitionName);
+    
+    const result = `
+      ${detachPartition};
+      ALTER TABLE ${this.tableName}
```
Comment (rathboma) [inline shared/src/lib/sql/change_builder/PostgresqlChangeBuilder.ts:49]: Escape!

**Comment by rathboma** [comment]
Excellent or PR overall, I think if you’d done your own code review you’d have caught the escaping issues :-).

use knex whenever possible to avoid having to manually escape.

**Comment by not-night-but** [comment]
@rathboma Changes made. A couple notes:
1. In PostgresqlChangeBuilder, this.tableName is a getter that automatically wraps the tableName, so we don't need to escape it
2. Also in that file, partitioning expressions are rawSql right now, my reasoning for that this is kinda like the raw filters for tables in TableTable, and we don't escape those at all. If we really don't like that I can probably do some wizardry to parse the expressions so that we just see either the low value and the high value, the list of values, or the modulus and remainder values. This was just the simplest solution.

File: apps/studio/src/lib/db/clients/postgresql.ts [hunk 12]
```diff
@@ -854,21 +854,26 @@ export async function getTableProperties(conn: HasPool, table: string, schema: s
...
   ] = await Promise.all([
     detailsPromise,
     listTableIndexes(conn, table, schema),
     getTableKeys(conn, "", table, schema),
     triggersPromise,
+    partitionsPromise,
     getTableOwner(conn, table, schema)
   ])
 
+  console.log('Partitions: ', partitions)
```
Comment (rathboma) [inline apps/studio/src/lib/db/clients/postgresql.ts:875]: Remove console logs. If you want to keep, make it a log.debug or log.imfo

**Review by rathboma — APPROVED** [review]
Looks good. Remove the console.log and merge it!

**Pull request merged**